Warn apps that use KVC on WKWebViewConfiguration about future breakage

bnham opened the pull request:
#### 8e951ee9cae383cff766c5830d7f1bd828e871e4
<pre>
Warn apps that use KVC on WKWebViewConfiguration about future breakage
<a href="https://bugs.webkit.org/show_bug.cgi?id=272358">https://bugs.webkit.org/show_bug.cgi?id=272358</a>
<a href="https://rdar.apple.com/125788604">rdar://125788604</a>

Reviewed by Chris Dumez.

We renamed some ivars in WKWebViewConfiguration in 276117@main. This broke some poorly behaved apps
that were accessing ivars on WKWebViewConfiguration via KVC. We&apos;ll temporarily allow these apps to
continue to function for backwards compat reasons when linking against an older SDK, but log a fault
when it happens. When linking against newer SDKs, Foundation will raise an exception about trying to
use KVC against a non-existent property for us.

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration setValue:forKey:]):
(-[WKWebViewConfiguration valueForKey:]):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/41802febe2f52ad6af8a146ca71c9fd1113ade51

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47160 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26339 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49803 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49843 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43209 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23796 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/38415 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47741 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23816 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40648 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19725 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21231 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41801 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5204 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/40426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/43536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42207 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51719 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/46646 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/22185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18555 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45708 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23462 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44716 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10398 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/24244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/54149 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23179 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/11111 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -293,6 +293,27 @@ - (id)copyWithZone:(NSZone *)zone
 
+- (void)setValue:(id)value forKey:(NSString *)key
+{
+    if (([key isEqualToString:@"allowUniversalAccessFromFileURLs"] || [key isEqualToString:@"_allowUniversalAccessFromFileURLs"]) && [value isKindOfClass:[NSNumber class]] && !linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::ThrowOnProblematicWebViewConfigurationKVCAccess)) {
```
I understand the check for `_allowUniversalAccessFromFileURLs` since this ivar used to exist and no longer does and we want to maintain backward compatibility. However, why the check for `allowUniversalAccessFromFileURLs`?

cdumez replied:
Also, doesn't this patch still allow setting other private ivars that still exist? Seems like if our new policy is to throw, we should do it consistently?

bnham replied:
After some discussion with other team members I'm going to change this patch slightly to block access to all ivars when linked against new SDKs.

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
```diff
@@ -293,6 +293,27 @@ - (id)copyWithZone:(NSZone *)zone
+
+- (id)valueForKey:(NSString *)key
+{
+    if (([key isEqualToString:@"allowUniversalAccessFromFileURLs"] || [key isEqualToString:@"_allowUniversalAccessFromFileURLs"]) && !linkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::ThrowOnProblematicWebViewConfigurationKVCAccess)) {
```
Same comment as above, why do we still allow apps to access other private ivars? Shouldn't we prevent this?

achristensen07 replied:
There aren't ivars any more.

cdumez replied:
Oh, I see. That's one way to prevent it :)

cdumez reviewed (approved):
Still curious why we need to check "allowUniversalAccessFromFileURLs" in addition to "_ allowUniversalAccessFromFileURLs" but LGTM given that there are no more ivars.

bnham committed:
Warn apps that use KVC on WKWebViewConfiguration about future breakage
https://bugs.webkit.org/show_bug.cgi?id=272358
rdar://125913705

Reviewed by Chris Dumez.

We renamed some ivars in WKWebViewConfiguration in 276117@main. This broke some poorly behaved apps
that were accessing ivars on WKWebViewConfiguration via KVC. We'll temporarily allow these apps to
continue to function for backwards compat reasons when linking against an older SDK, but log a fault
when it happens. When linking against newer SDKs, Foundation will raise an exception about trying to
use KVC against a non-existent property for us.

* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration setValue:forKey:]):
(-[WKWebViewConfiguration valueForKey:]):

Canonical link: https://commits.webkit.org/277266@main
